### PR TITLE
Update config

### PR DESCRIPTION
**Pull request by ilan-gold**
@manzt Is this what you were looking for?  We only have one image so I put all of this in one "image."

### CODE REVIEW

**Comment by manzt** [comment]
I meant we need to transfer the urls for the images your had before in the viv-benchmark repo. The config that's currently in place served to prototype the benchmark script.

We need to:

1.) Configure all the endpoints we want to test (sources).
2.) Configure the different regions (including zoom levels for the different images)
3.) Actually run the benchmark, probably not on my local machine and a server.

PS: I keep getting `Failed to load resource: net::ERR_CERT_INVALID` with any of the new URLs. I'm not sure what this error is, but I think it is related to something you brought up last week. Hopefully you can figure out how to configure the right certs in pupeteer.

**Comment by manzt** [comment]
I also can't figure out if the `--disable-http2` flag is working. I've tried running it not in headless mode, and can't find in the requests. It would be great to some how assess the protocol from the fetch request itself. do you know if that's possible?

```javascript
const browser = await puppeteer.launch({ headless: false });
``

**Comment by ilan-gold** [comment]
Yes you can do that if you can get your hands on the HAR file or if you can look at the network tabs of the browser (and check protocol).  I don't think we can really deduce this outside of an explicit setting, though (i.e deducing it from the fetch itself).  We can just configure the script to take `http1` or `http2` as a configuration parameter and use different URLs.  We have the servers.

**Comment by ilan-gold** [comment]
That last comment was a little mushy so I guess the question is: Do you want me to set up the `http1` and `http2` as separate URLs or do you want me to try to get it to work with the flag?

**Comment by ilan-gold** [comment]
P.S What other images did you want to use?  The servers only have the Vanderbilt ones on them and generating other ones takes time.

**Comment by manzt** [comment]
> That last comment was a little mushy so I guess the question is: Do you want me to set up the http1 and http2 as separate URLs or do you want me to try to get it to work with the flag?

I think I'd prefer that it's just a flag. That means we can use one URL and run the benchmark on the same config with different flags:

```javascript
run('http1.csv', { args: ['--disable-http2'] });
run('http2.csv');
```

vs

```javascript
run('all_data.csv') // here http1/2 is encoded in the source urls
```

**Comment by manzt** [comment]
The only other image I had in mind perhaps the one for the tutorial. It's a lower bit-depth, but publicly available under CC4.

https://github.com/hms-dbmi/viv/tree/master/tutorial

**Comment by ilan-gold** [comment]
@manzt I can confirm that with `headless: false,` you can see that `--disable-http2` works as expected (seen in the opened the network tab), although I am skeptical since I saw this previously and then it suddenly did not work.  I had trouble intercepting requests to view the http version (i.e in headless mode).   Nonetheless, I can see that for now it works in `headless: false`.  I think we can add a `--auto-open-devtools-for-tabs` so you may check this fact while the tests are running.  I think `headless: false, args: ["--auto-open-devtools-for-tabs"]` as a baseline for both tests would be good.

**Comment by ilan-gold** [comment]
> @manzt I can confirm that with `headless: false,` you can see that `--disable-http2` works as expected (seen in the opened the network tab), although I am skeptical since I saw this previously and then it suddenly did not work. I had trouble intercepting requests to view the http version (i.e in headless mode). Nonetheless, I can see that for now it works in `headless: false`. I think we can add a `--auto-open-devtools-for-tabs` so you may check this fact while the tests are running. I think `headless: false, args: ["--auto-open-devtools-for-tabs"]` as a baseline for both tests would be good.

FWIW this already appears to be very fickle, as I observed previously - I just tested it out and it was `http1`. Can you test this once or twice?  In addition I noticed that we aren't using a queue - is there a reason for that?

**Comment by ilan-gold** [comment]
Ah @manzt it looks like the funcions `run` are async....so maybe that was it since we do not await them.

**Comment by manzt** [comment]
It shouldn't be an issue to run the scripts concurrently, but for monitoring this would be good. We could just write separate scripts or a cli:

```bash
$ ./benchmark.js chrome_http1.csv --disable-http2 #<other flags to pass to args>
$ ./benchmark.js chrome_http2.csv
```

**Comment by ilan-gold** [comment]
@manzt Do we want to use a queue?

**Comment by manzt** [comment]
@ilan-gold see #4. I'd recommend this rather than the commit above fro separating http1/2. Easier to add different flags, without editing `benchmark.js`

**Comment by manzt** [comment]
> @manzt Do we want to use a queue?

For what part? I think this might be reasonable in the index.js: 

```javascript
const times = await Promise.all(p); // await all requests
```

I _think_ in deckgl the max requests are set to `10` for the tile layer. Take a look at `p-queue` if this is of interest:

```javascript
const queue = new PQueue({ concurrency: 10 });
for (const t of tileCoords) {
	queue.add(() => timeGetTile(loader, { ...t, loaderSelection }));
}
// guarantees that all work on queue has finished
await queue.onIdle();
```

**Comment by ilan-gold** [comment]
@manzt At lower zoom levels, you can end up requesting 20 or 30 tiles at once, hence the queue, but I guess that is a relatively small number.  

Do you want me to remove the parts of my code that handled the separate processes here?  The only worry I have about #4 and going headless is that I don't trust the headless with http2 connection (when I was intercepting, they looked like HTTP1 but also intercepting them could cause that behavior).

**Comment by manzt** [comment]
Why don't we add the concurrency limit as described above. It seems more like the behavior that `deck.gl` would emulate.

**Comment by manzt** [comment]
> Do you want me to remove the parts of my code that handled the separate processes here? The only worry I have about #4 and going headless is that I don't trust the headless with http2 connection (when I was intercepting, they looked like HTTP1 but also intercepting them could cause that behavior).

Yeah why don't we just make #4 handle this and you can merge this branch with master after we merge that.

File: src/index.js [hunk 1]
```diff
@@ -27,7 +28,13 @@ async function timeRegions({ file, sources, regions }, iter) {
...
           });
-          const p = tileCoords.map(t => timeGetTile(loader, { ...t, loaderSelection }));
+          const queue = new PQueue({ concurrency: 10 });
+          const p = []
+          for (const t of tileCoords) {
+            p.push(queue.add(() => timeGetTile(loader, { ...t, loaderSelection })));
+          }
+          // guarantees that all work on queue has finished
+          await queue.onIdle();
           const times = await Promise.all(p); // await all requests
```
Comment (manzt) [inline src/index.js]:
```suggestion
          const p = tileCoords.map(t => timeGetTile(loader, { ...t, loaderSelection }));
          const queue = new PQueue({ concurrency: 10 });
          const records = []
          for (const t of tileCoords) {
            queue.add(async () => {
            	const [start, end] = await timeGetTile(loader, { ...t, loaderSelection });
				const { x, y, z } = t;
				const record = {
                  iter,
                  file,
                  format,
                  compression,
                  // The zoom level is the z, not the provided one.
                  zoom: -z,
                  tileSize,
                  regionId: id,
                  yCoord,
                  xCoord,
                  height,
                  width,
                  numChannels: n,
                  tileId: `${x}-${y}-${z}`,
                  startTime: start,
                  endTime: end,
                };
                records.push(record);
            );
          }
          // guarantees that all work on queue has finished
          await queue.onIdle();

Comment (manzt) [reply src/index.js]: I don't think queue.add() returns anything, so you need to capture the records as they are executed.

Comment (manzt) [reply src/index.js]: and then just await the queue when it's finished

Comment (ilan-gold) [reply src/index.js]: Sorry, it seemed from the documentation that it returned a promise, but this is cleaner.

Comment (manzt) [reply src/index.js]: Ah, it does return a Promise, but I'm not sure if it resolves to a value. So we need to collect the data somehow.

Comment (ilan-gold) [reply src/index.js]: It seemed like it was just resolving to the return value, but this is cleaner nonetheless.  I updated it.

**Comment by ilan-gold** [comment]
@manzt Tested locally - should be ready to merge.  I can run this overnight.

File: README.md [hunk 2]
```diff
@@ -1,4 +1,5 @@
 ```bash
 $ npm install && npm run build
-$ node benchmark.js
+$ node benchmark.js --use-http2
+$ node benchmark.js --use-http1
```
Comment (manzt) [inline README.md]:
```suggestion
$ npm run bench # runs benchmark
$ ./benchmark.js outfile.csv [--no-headless] [--ignore-https-errors] [chrome-flags]
```

**Review by manzt — APPROVED** [review]
Great, thanks! Looks good to run to me.

**Pull request merged**